pkg/services/ngalert/state/manager_private_test.go

```diff
@@ -738,6 +739,262 @@ func TestProcessEvalResults_StateTransitions(t *testing.T) {
 				},
 			},
 		},
+		{
+			desc:      "t1[{}:alerting] t2[{}:recovering] t3[{}:recovering] t4[{}:normal] and 'keep_firing_for'>0  at t2,t3.t4",
+			alertRule: baseRuleWith(ngmodels.RuleMuts.WithKeepFiringForNTimes(2), ngmodels.RuleMuts.WithFor(0)),
+			results: map[time.Time]eval.Results{
+				t1: {
+					newResult(eval.WithState(eval.Alerting)),
+				},
+				t2: {
+					newResult(eval.WithState(eval.Normal)),
+				},
+				t3: {
+					newResult(eval.WithState(eval.Normal)),
+				},
+				t4: {
+					newResult(eval.WithState(eval.Normal)),
+				},
+			},
+			expectedTransitions: map[time.Time][]StateTransition{
+				t1: {
+					{
+						PreviousState: eval.Normal,
+						State: &State{
+							Labels:             labels["system + rule"],
```

Comment:
let's use `"system + rule + labels1"` everywhere in these tests. `"system + rule"` usually happens when no-data or classic_condition